# <font color= white> 기말고사 - 통계분석론 2020-2학기

<font color="white"> 
    
- **각 문항에 답을 하기 위해 사용된 그래프, 표, 통계량 등을 반드시 모두 제시하세요.**
- **아래의 문제에 답을 채워넣은 뒤 모든 결과물이 프린트 된 상태에서 `File>Download as>HTML(.html)` 형식으로 저장하여 업로드하세요.**
- **모든 가설검정은 유의수준($\alpha$) 0.05 하에서 진행하세요.**

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
from scipy import stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats import weightstats
from statsmodels.stats.outliers_influence import OLSInfluence


1. 한 연구자는 우울증 치료요법의 효과를 측정하였다(depression.csv). 종속변수 y 는 측정된 치료요법의 효과(클수록 효과가 좋음)을 나타내고, age는 환자의 나이, TRT(A, B, C)는 세 가지 종류의 치료요법을 나타낸다. Y를 age와 TRT로 설명하는 회귀식을 적합하고 아래의 질문에 답하시오.

(1)	추정된 회귀식을 쓰고 이 회귀식이 종속변수를 설명하는데 유의한지 가설검정을 진행하시오. (귀무가설, 대립가설 명시)

In [1]:
dep = pd.read_csv(r"C:/Users/MSI/Desktop/depression.csv")

In [2]:
dep.head()

y  age TRT
0  56   21   A
1  41   23   B
2  40   30   B
3  28   19   C
4  55   28   A

In [5]:
model = smf.ols('y ~ age + TRT', data = dep).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.784
Model:                            OLS   Adj. R-squared:                  0.764
Method:                 Least Squares   F-statistic:                     38.71
Date:                Wed, 16 Dec 2020   Prob (F-statistic):           9.29e-11
Time:                        18:04:52   Log-Likelihood:                -113.68
No. Observations:                  36   AIC:                             235.4
Df Residuals:                      32   BIC:                             241.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     32.5433      3.581      9.088      0.000      25.249      39.838
TRT[T.B]      -9.8076      2.465     -3.979      0.000     -14.828      -4.787
TRT[T.C]     -10.2528      2.465     -4.159      0.000     -15.275      -5.231
age            0.6645      0.070      9.522      0.000       0.522       0.807
==============================================================================
Omnibus:                        1.212   Durbin-Watson:                   2.214
Prob(Omnibus):                  0.546   Jarque-Bera (JB):                1.196
Skew:                          -0.347   Prob(JB):                        0.550
Kurtosis:                       2.438   Cond. No.                         184.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

H0: There is no effectiveness in the depression therapies.
H1: Depression therapy is effective.

Because the p-value of each coefficient is less than 0.05, each regression coefficient is significant. This can be said to be effective in therapy.


(2) 추정된 회귀식은 Y의 변동량의 몇 %를 설명하는가?

R2: 0.784.
therefore, The estimated regression expression accounts for 78.40% of the variation in Y.

(3) age에 대한 회귀계수가 유의한지 가설검정을 진행하시오. (귀무가설, 대립가설 제시)

H0: Age is not significant for the regression.
H1: Age is significant for the regression.

Because the p-value of Age is less than 0.05, the variable is significant. But it should also be taken with carefulness. 유의하다

(4) Age와 trt사이에 교호작용(상호작용)이 존재하는가? 가설검정을 통해 확인하시오.

In [7]:
model2 = smf.ols('y ~ age + TRT + age*TRT', data = dep).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.914
Model:                            OLS   Adj. R-squared:                  0.900
Method:                 Least Squares   F-statistic:                     64.04
Date:                Wed, 16 Dec 2020   Prob (F-statistic):           4.26e-15
Time:                        18:14:39   Log-Likelihood:                -97.024
No. Observations:                  36   AIC:                             206.0
Df Residuals:                      30   BIC:                             215.5
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       47.5156      3.825     12.422      0.000      39.703      55.328
TRT[T.B]       -18.5974      5.416     -3.434      0.002     -29.658      -7.537
TRT[T.C]       -41.3042      5.085     -8.124      0.000     -51.688     -30.920
age              0.3305      0.081      4.056      0.000       0.164       0.497
age:TRT[T.B]     0.1932      0.117      1.657      0.108      -0.045       0.431
age:TRT[T.C]     0.7029      0.109      6.451      0.000       0.480       0.925
==============================================================================
Omnibus:                        2.593   Durbin-Watson:                   1.633
Prob(Omnibus):                  0.273   Jarque-Bera (JB):                1.475
Skew:                          -0.183   Prob(JB):                        0.478
Kurtosis:                       2.079   Cond. No.                         577.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

H0: There is no iteraction 
H1: There is iteraction

because the p value for Age and tTRT-B is larger than 0.05,the null hypothesis cannot be rejected. In this case there is no apparent interaction.

Nonetheless, the p value for age and TRT - C is less than 0.05, so we reject the null hypothesis and conclude that there is an iteraction effect

(5) 치료요법 A,B,C 중 어느 것이 효과가 좋다고 할 수 있는가? (4)의 결과를 사용하여 서술하시오.

When "age" is fixed, on average it decreases by 18.59 when "TRT-B" is used relative to "TRT-A".

The use of "TRT-C" is less effective than using "TRTA" because the interaction with "age" results in a larger reduction of 41.30. Therefore using treatment A is most effective (since is the reference level and the other two treatments are being compared to this one).

2\. 한 연구자는 대학원 지원자의 GRE(graduate record exam; gre) 점수, 학점(gpa), 학부 졸업대학의 랭킹(rank)이 합격여부(admit=1:합격, 0:불합격)에 미치는 영향을 알아보고자 하였다. 데이터는 `admission.csv`에 저장되어 있다. `admit`을 종속변수로, 나머지 변수를 설명변수로 사용하여 로지스틱 회귀모형을 적합하고 그 결과를 사용하여 아래 문항에 답하시오. 

In [11]:
gre = pd.read_csv(r"C:/Users/MSI/Desktop/admission.csv")

In [12]:
gre.head()

admit  gre   gpa  rank
0      0  380  3.61     3
1      1  660  3.67     3
2      1  800  4.00     1
3      1  640  3.19     4
4      0  520  2.93     4

In [15]:
model = smf.glm('admit~ gre+gpa+ rank', data = gre, family = sm.families.Binomial()).fit()
model.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

(1) 이미 GRE점수를 받아놓은 한 지원자가 있다. 현재 학점이 3.5인데 계절학기 재수강을 하면 학점을 3.7로 올릴 수 있다. 만일 학점을 3.7로 올린다면 이 학생이 합격할 odds는 몇 % 상승하는가? 

In [18]:
#GRE 3.7-3.5 = 0.2
np.exp(model.params[1]*0.2)

1.00045889716199

R: A 0.2 increase in **gre** score increases the acceptable odds by 0.045%.

####  (I think the question was directed to using the GPA because of the 3.7 and 3.5 values. In this case the solution is that a 0.2 increase in the GPA increases the acceptable odds by 22.95%  

#### code:       np.exp(model.params[2]*0.2))

(2) 만일 학점과 GRE점수가 동일한 두 학생이 각기 rank2 의 대학과 rank3의 대학을 졸업했다고 하자. Rank3에 속한 대학을 졸업한 학생에 비해 rank2에 속한 대학을 졸업한 학생의 합격 odds는 몇 % 증가하는가? 

In [19]:
#odds Rank 2 university (with gre and gpa constant)
model2 = smf.glm('admit~ gre+gpa+ C(rank)', data = gre, family = sm.families.Binomial()).fit()

In [20]:
model2.params

Intercept      -3.989979
C(rank)[T.2]   -0.675443
C(rank)[T.3]   -1.340204
C(rank)[T.4]   -1.551464
gre             0.002264
gpa             0.804038
dtype: float64

In [21]:
np.exp(model2.params[1])

0.5089309509281347

In [22]:
#Rank 2 university porobability
(1-np.exp(model2.params[1]))*100

49.10690490718653

In [24]:
#odds Rank 3 university (with gre and gpa constant)
np.exp(model2.params[2])

0.26179227938011124

In [25]:
#Rank 3 university porobability
(1-np.exp(model2.params[2]))*100

73.82077206198888

The probability of entering Rank 2 university (mantaining other values constant) is of 49.11% and of entering rank 3 university is of 73.82%.

3\. 한 연구자는 혈압(blood pressure)에 영향을 주는 요인을 파악하기 위해 20명으로부터 아래의 변수를 측정하였다. (`bloodpress.csv`)

|  변수명      |   내용                                |
|--------|-----------------------------------|
| BP     | Blood pressure in mm Hg           |
| Age    | age in years                      |
| Weight | weight in kg                      |
| BSA    | body surface area in sq m         |
| Dur    | duration of hypertension in years |
| Pulse  | basal pulse in beats per minute   |
| Stress | stress index                      |

In [26]:
bp = pd.read_csv(r"C:/Users/MSI/Desktop/bloodpress.csv")

(1) BP를 종속변수로 하고 나머지 6개의 변수를 설명변수로 하는 선형회귀모형을 적합하고 VIF를 통해 다중공선성이 존재하는지 판단하시오. 

In [32]:
formula = smf.ols('BP~Age +Weight+BSA+Dur+Pulse+Stress', data = bp).fit()
formula.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     BP   R-squared:                       0.996
Model:                            OLS   Adj. R-squared:                  0.994
Method:                 Least Squares   F-statistic:                     560.6
Date:                Wed, 16 Dec 2020   Prob (F-statistic):           6.40e-15
Time:                        18:39:33   Log-Likelihood:                -6.1033
No. Observations:                  20   AIC:                             26.21
Df Residuals:                      13   BIC:                             33.18
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -12.8705      2.557     -5.034      0.000     -18.394      -7.347
Age            0.7033      0.050     14.177      0.000       0.596       0.810
Weight         0.9699      0.063     15.369      0.000       0.834       1.106
BSA            3.7765      1.580      2.390      0.033       0.363       7.190
Dur            0.0684      0.048      1.412      0.182      -0.036       0.173
Pulse         -0.0845      0.052     -1.637      0.126      -0.196       0.027
Stress         0.0056      0.003      1.633      0.126      -0.002       0.013
==============================================================================
Omnibus:                        7.219   Durbin-Watson:                   2.249
Prob(Omnibus):                  0.027   Jarque-Bera (JB):                4.643
Skew:                          -1.050   Prob(JB):                       0.0981
Kurtosis:                       4.076   Cond. No.                     4.02e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.02e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [33]:
import patsy
y, X = patsy.dmatrices('BP~Age +Weight+BSA+Dur+Pulse+Stress', data = bp, return_type = "matrix" )
from statsmodels.stats.outliers_influence import variance_inflation_factor
for i, x in enumerate(X.design_info.column_names):
    print( 'VIF of', x, ':', variance_inflation_factor(X, i))

VIF of Intercept : 788.3076923918795
VIF of Age : 1.7628067217672172
VIF of Weight : 8.41703502963307
VIF of BSA : 5.328751470118884
VIF of Dur : 1.2373094205198356
VIF of Pulse : 4.413575165597264
VIF of Stress : 1.8348453242645855


For this case, the variables "dur", "Pulse", "Stress" are no significative for the regression of the model. 

if VIF is greater than 5, then the explanatory variable is highly collinear with the other explanatory variables. In this case, Weight and BSA present multicolineality. With this, the parameter estimates  will have large standard errors.

(2) 설명변수 중 Age는 반드시 모형에 포함되어야 한다고 가정하자. Age와 또 다른 하나의 변수를 포함한 **설명변수가 2개**인 모든 모형들을 비교하여 BIC를 기준으로 최적의 모형을 찾고 적합된 결과물을 프린트하시오. 

In [46]:
import itertools
import time
import statsmodels.api as sm

def processSubset(feature_set):
    # Fit model on feature_set and calculate RSS
#     X_model = sm.add_constant(X[list(feature_set)])
    X_model = X[list(('Intercept',)+feature_set)]
    model = sm.OLS(y, X_model)
    regr = model.fit()
    RSS = (regr.resid ** 2).sum()
    return {"model":regr, "RSS":RSS, "AIC": regr.aic, 
            "BIC": regr.bic, "adj_r2": regr.rsquared_adj,
           "r2":regr.rsquared }

def getBest(k, best = "AIC"):
    
    tic = time.time()   
    results = []
    
    for combo in itertools.combinations(X.columns[1:], k):
        results.append(processSubset(combo))
    
    # Wrap everything up in a nice dataframe
    models = pd.DataFrame(results)
    
    # Choose the model with the highest RSS
    if best=="adj_r2" or best=="r2":
        best_model = models.loc[models[best].argmax()]
    else: 
        best_model = models.loc[models[best].argmin()]
    
    toc = time.time()
    print("Processed", models.shape[0], "models on", k, "predictors in", (toc-tic), "seconds.")
    
    # Return the best model, along with some other useful information about the model
    return best_model

In [47]:
import patsy
y, X = patsy.dmatrices('BP~Age +BSA', data = bp, return_type = 'dataframe')

In [50]:
models1= pd.DataFrame(columns=["model", "RSS","AIC","BIC","adj_r2", "r2"])

for i in range(1,X.shape[1]-1):
    models1.loc[i] = getBest(i, best = "BIC")

Processed 2 models on 1 predictors in 0.00500035285949707 seconds.


In [51]:
models1

model         RSS        AIC  \
1  <statsmodels.regression.linear_model.Regressio...  140.142123  99.696037   

          BIC    adj_r2        r2  
1  101.687502  0.735843  0.749746

In [64]:
import patsy
y, X = patsy.dmatrices('BP~ Age+ Weight ', data = bp, return_type = 'dataframe')

In [65]:
models2= pd.DataFrame(columns=["model", "RSS","AIC","BIC","adj_r2", "r2"])

for i in range(1,X.shape[1]-1):
    models2.loc[i] = getBest(i, best = "BIC")

Processed 2 models on 1 predictors in 0.006001710891723633 seconds.


In [67]:
models2

model        RSS        AIC  \
1  <statsmodels.regression.linear_model.Regressio...  54.528016  80.817188   

         BIC    adj_r2        r2  
1  82.808653  0.897219  0.902629

In [56]:
import patsy
y, X = patsy.dmatrices('BP~ Age+ Dur ', data = bp, return_type = 'dataframe')

In [57]:
models3= pd.DataFrame(columns=["model", "RSS","AIC","BIC","adj_r2", "r2"])

for i in range(1,X.shape[1]-1):
    models3.loc[i] = getBest(i, best = "BIC")

Processed 2 models on 1 predictors in 0.004999399185180664 seconds.


In [59]:
models3

model         RSS         AIC  \
1  <statsmodels.regression.linear_model.Regressio...  316.734007  116.004142   

          BIC    adj_r2        r2  
1  117.995607  0.402982  0.434404

In [60]:
import patsy
y, X = patsy.dmatrices('BP~ Age+ Pulse ', data = bp, return_type = 'dataframe')

In [61]:
models4= pd.DataFrame(columns=["model", "RSS","AIC","BIC","adj_r2", "r2"])

for i in range(1,X.shape[1]-1):
    models4.loc[i] = getBest(i, best = "BIC")

Processed 2 models on 1 predictors in 0.005999088287353516 seconds.


In [63]:
models4

model         RSS         AIC  \
1  <statsmodels.regression.linear_model.Regressio...  268.555313  112.704034   

          BIC    adj_r2        r2  
1  114.695498  0.493795  0.520437

In [68]:
import patsy
y, X = patsy.dmatrices('BP~ Age+ Pulse ', data = bp, return_type = 'dataframe')

In [69]:
models5= pd.DataFrame(columns=["model", "RSS","AIC","BIC","adj_r2", "r2"])

for i in range(1,X.shape[1]-1):
    models5.loc[i] = getBest(i, best = "BIC")

Processed 2 models on 1 predictors in 0.005035877227783203 seconds.


In [71]:
models5

model         RSS         AIC  \
1  <statsmodels.regression.linear_model.Regressio...  268.555313  112.704034   

          BIC    adj_r2        r2  
1  114.695498  0.493795  0.520437

The smallest BIC value is for the Age and Weight pair with a 82.81 BIC value

In [73]:
model = smf.ols('BP~ Age + Weight', data = bp).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     BP   R-squared:                       0.991
Model:                            OLS   Adj. R-squared:                  0.990
Method:                 Least Squares   F-statistic:                     978.2
Date:                Wed, 16 Dec 2020   Prob (F-statistic):           2.81e-18
Time:                        19:26:09   Log-Likelihood:                -14.157
No. Observations:                  20   AIC:                             34.31
Df Residuals:                      17   BIC:                             37.30
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -16.5794      3.007     -5.513      0.000     -22.925     -10.234
Age            0.7083      0.054     13.235      0.000       0.595       0.821
Weight         1.0330      0.031     33.154      0.000       0.967       1.099
==============================================================================
Omnibus:                        0.989   Durbin-Watson:                   1.688
Prob(Omnibus):                  0.610   Jarque-Bera (JB):                0.768
Skew:                           0.101   Prob(JB):                        0.681
Kurtosis:                       2.061   Cond. No.                     2.65e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.65e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [74]:
formula2 =('BP~ Age + Weight')

In [75]:
import patsy
y, X = patsy.dmatrices(formula2, bp, return_type='matrix')
from statsmodels.stats.outliers_influence import variance_inflation_factor
for i, x in enumerate(X.design_info.column_names):
    print( 'VIF of', x, ':', variance_inflation_factor(X, i))

VIF of Intercept : 637.4961503521716
VIF of Age : 1.1989450542463735
VIF of Weight : 1.1989450542463729
